Using user's account language to set campaign params

blisteringherb opened the pull request:
- When user's sign up for a campaign, they will now receive their signup
  emails with copy in their language of choice.
- The user's language should always be set appropriately either when they create an account or when they change their own language settings.

Fixes #5091 and #4956

Commit:
Using user's account language to set campaign params

* When user's sign up for a campaign, they will now receive their signup
  emails with copy in their language of choice.

angaither commented:
code looks :ok: 
but I don't think we are allowing users to change their own lang? what if this isn't set?

blisteringherb commented:
The language by default when a user creates an account will be 'en'. We'll need a hook_user_insert to make sure the language gets set properly based on the Fastly headers, so the user's language should always be set properly. If the hook_user_insert malfunctions for some reason then the user will get content in 'en'. We may want to log when the user's language setting is different than what we get from Fastly.

itsjoekent commented:
Yeah this was the point I raised yesterday @angaither 

https://dosomething.slack.com/archives/global/p1442416380000776

We decided it's safe to assume when a user registers they've always have the correct language set

angaither commented:
...right but what about users we have already w/o a lang?

itsjoekent commented:
what users dont have a language set? dont they have to have at least english set?

blisteringherb commented:
@angaither @deadlybutter Hm, it looks like in the db the language column is empty or und for all but about 6 users. Should we run a script now before we turn on global to set all of those users' language to 'en'?  Is there any advantage to keeping the language empty or 'und' for a user?

itsjoekent commented:
I don't see any advantage, I think we should put everyone on the same system. @angaither ?

angaither commented:
@blisteringherb rather than writing a script to handle this, I have sort of done it here
https://github.com/DoSomething/phoenix/blob/dev/lib/modules/dosomething/dosomething_global/dosomething_global.module#L116

blisteringherb commented:
@angaither That function accounts for users with global admin accounts but not others. We could expand that function or write some other helpers to return the language based on logic, but I think it would be more performant if we make sure the user's language is set properly in hook_user_insert and then generally use the user's language whenever we need to show them something.

angaither commented:
@blisteringherb that sounds good to me, all I meant was we should have a default of english if the user doesn't have a language set on insert (which I think is the case for most of our users now)

itsjoekent commented:
according to morgans comment above, only 6 users don't have a language set. can we set it for them & assume from thereon out all users will have a language set?

angaither commented:
@deadlybutter, he said all but 6 have a lang set. 
I just ran the query ~20 k have something set
the other 3 million have nothing or `und` set

itsjoekent commented:
OH. well that changes things. 

EDIT: so yeah, we should probably make a helper function which we call to grab a users language that can provide a default fallback. (should this be global-en, or en?)

mikefantini commented:
since the majority of our current users are :us: and the ratio of new users will most likely stay majority :us: for a while I would say default should be en.

mshmsh5000 commented:
:+1:. Because of #5130, the system's default language may become Global English. Makes sense for users still to default to US English, though.

deezone commented:
@blisteringherb @mikefantini There might need to be clarification on this issue as my understanding is a campaign sign up should trigger a transactional email in the **language of the campaign** NOT in the language of the user (as the issue description is suggesting).

Looking at the code changes in this PR it looks like you are using the language of the campaign. Just wanted to clarify as the bulk of the comments on this issue are discussing the user's language preference at the time of signup.

mikefantini commented:
@blisteringherb @DeeZone the languages of the emails should be as follows:

| Type of Email | Language |
| --- | --- |
| User Registration | Language of user |
| PW Reset | Language of user |
| Campaign signup | Language of campaign |
| Campaign reportback | Language of campaign |

Let me know if any further clarification is needed.

blisteringherb commented:
@DeeZone @mikefantini I'm unsure what you mean by "language of campaign". The user's language preference or the language from the URL are the only languages that we can use. A campaign will have multiple languages.

mikefantini commented:
@blisteringherb take the following case:
- Brazilian user
- signs up for node 123
- node 123 has a US English translation published and a Global English translation published but has no BR-PT translation published
- in this case even though the user's language setting would be PT-BR, because she signed up for a Global English translation of the campaign, her transactional email should be in Global English, not BR-PT
- user then signs up for node 124
- node 124 has a US English translation published, a Global English translation published, and a BR-PT translation published
- in this case, the user would be seeing the BR-PT translation of the campaign and would then get a transactional email for node 124 in BR-PT

does that make sense/seem doable @blisteringherb?

Commit:
Adding a helper function to get language for node or for account

* The helper function can be called from anywhere to get the appropriate
  language depending on the context. Pass in a node if determining the
  lang for user with a node. This should help keep all references to language
  determining logic in one place.

* Refactoring campaign signup to use language helper function

Commit:
Fixing merge conflict, adding lang helper func

angaither commented on lib/modules/dosomething/dosomething_global/dosomething_global.module:
```diff
@@ -221,4 +221,37 @@ function dosomething_global_convert_country_to_language($country) {
+ *
+ */
+function dosomething_global_get_language($account, stdClass $node = NULL) {
+  if ($node = $options['node']) {
```
`$options['node']` isn't getting passed in?

Commit:
Fixing wrong reference to $node object

angaither commented on lib/modules/dosomething/dosomething_global/dosomething_global.module:
```diff
@@ -221,4 +221,37 @@ function dosomething_global_convert_country_to_language($country) {
+    }
+  }
+  elseif ($language = $account->language) {
+    return $language;
```
shouldn't this say so you aren't returning in the middle?

``` php
elseif (isset($account->language)) {
 $language = $account->language
}
```

blisteringherb replied:
Is there a reason to avoid returning here?

mshmsh5000 replied:
This is an assignment, not a comparison. `if ($language = $account->language)` will always be `true`.

If the intended logic is "if the current language is equal to the account language, then just return the language here and skip the last few lines," that makes sense to me.

angaither replied:
:+1: ya agree, I think I just misunderstood that

Commit:
Missing brace
